### PR TITLE
[App Service] `az functionapp create`: Add `--zone-redundant` parameter to support zone redundant for Functions Flex SKU

### PR DESCRIPTION
**Pull request by khkh-ms**
**Related command**
az functionapp create

**Description**
Adding the support of --zone-redundant flag in `az functionapp create` for Flex SKU. 

**Testing Guide**
az functionapp create -g [rg] -n [name] -s [storage-account-name] --functions-version 4 --runtime python --runtime-version 3.11 --flexconsumption-location eastasia --zone-redundant

**History Notes**

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 1]
```diff
@@ -4807,6 +4819,12 @@ def create_functionapp(cmd, resource_group_name, name, storage_account, plan=Non
     SiteConfig, NameValuePair, DaprConfig, ResourceConfig = cmd.get_models('SiteConfig', 'NameValuePair',
                                                                            'DaprConfig', 'ResourceConfig')
 
+    if flexconsumption_location is None:
+        if zone_redundant:
+            raise ArgumentUsageError(
+                '--zone-redundant is only valid input for Azure Functions on the Flex Consumption plan. '
```
[the remark this answers is not in the document]
Comment (khkh-ms) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:4825]: You are right. I will change the wording. Thanks!

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 2]
```diff
@@ -822,6 +826,8 @@ def load_arguments(self, _):
         c.argument('deployment_storage_auth_value', options_list=['--deployment-storage-auth-value', '--dsav'], help="The deployment storage account authentication value. For the user-assigned managed identity authentication type, "
                    "this should be the user assigned identity resource id. For the storage account connection string authentication type, this should be the name of the app setting that will contain the storage account connection "
                    "string. For the system assigned managed-identity authentication type, this parameter is not applicable and should be left empty.", is_preview=True)
+        c.argument('zone_redundant', arg_type=get_three_state_flag(),
+                   help='Enable zone redundancy for high availability. Applies to Flex Consumption only.', is_preview=True)
```
Comment (sidkri) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:830]: "Applies to Flex Consumption SKU only."

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 3]
```diff
@@ -4752,6 +4756,13 @@ def is_exactly_one_true(*args):
     return found
 
 
+def list_flexconsumption_zone_redundant_locations(cmd):
+    client = web_client_factory(cmd.cli_ctx)
+    regions = client.list_geo_regions(sku="FlexConsumption")
```
Comment (sidkri) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4761]: What happens when this API call fails?  Does the operation fail or the method return no locations?

**Review by sidkri — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (khkh-ms) [reply src/azure-cli/azure/cli/command_modules/appservice/_params.py]: Everything related is flex is in preview mode. When we will remove the preview parameters from all Flex related flag, the preview from --zone_redundant will be removed at that time.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 3 above)
Comment (khkh-ms) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:4761]: AZ and SDK surfaces all API errors to the users.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 4]
```diff
@@ -5133,14 +5151,23 @@ def create_functionapp(cmd, resource_group_name, name, storage_account, plan=Non
         create_app_insights = True
 
     if flexconsumption_location is not None:
+        if zone_redundant:
+            zone_redundant_locations = list_flexconsumption_zone_redundant_locations(cmd)
+            zone_redundant_location = next((loc for loc in zone_redundant_locations
+                                            if loc['name'].lower() == flexconsumption_location.lower()), None)
+            if zone_redundant_location is None:
+                raise ValidationError("Location doesn't support zone redundancy in Flex Consumption. "
```
[the remark this answers is not in the document]
Comment (khkh-ms) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py]: Addressed it.

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py (hunk 2 above)
Comment (khkh-ms) [reply src/azure-cli/azure/cli/command_modules/appservice/_params.py]: Done

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (khkh-ms) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py]: Done

**Review by sidkri — APPROVED** [review]

**Comment by yanzhudd** [comment]
please add test for the newly added parameter to verify if it can meet your expectations.

**Comment by khkh-ms** [comment]
Added the tests.

**Comment by yonzhan** [comment]
Please fix CI issues

**Comment by khkh-ms** [comment]
It is not released to the PR. All the recent PRs are failing because the same failure. Please check and fix the pipeline if possible.

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**